src/librustc_mir/interpret/operand.rs

```diff
@@ -444,15 +443,12 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         place: mir::Place<'tcx>,
         layout: Option<TyAndLayout<'tcx>>,
     ) -> InterpResult<'tcx, OpTy<'tcx, M::PointerTag>> {
-        let base_op = match place.local {
-            mir::RETURN_PLACE => throw_ub!(ReadFromReturnPlace),
-            local => {
-                // Do not use the layout passed in as argument if the base we are looking at
-                // here is not the entire place.
-                let layout = if place.projection.is_empty() { layout } else { None };
-
-                self.access_local(self.frame(), local, layout)?
-            }
+        let base_op = {
+            // Do not use the layout passed in as argument if the base we are looking at
+            // here is not the entire place.
+            let layout = if place.projection.is_empty() { layout } else { None };
+
+            self.access_local(self.frame(), place.local, layout)?
```

Comment:
Is that really the only place where we treat the return place separately? I seem to recall we did some weird "NULL pointer place" things at some point, but maybe we got rid of this...

Reply:
Yeah I'm still hitting some other miri issues with #71003. The current issue is this ICE:

> error: internal compiler error: src/librustc_mir/interpret/eval_context.rs:142: The type checker should prevent reading from a never-written local

Apparently the return place is never marked as `Live`? It definitely gets initialized properly before being read. You can see the NRVO transformation here: https://gist.github.com/jonas-schievink/6be82985b216f4fa76df43fa890df0f1

Reply:
The return place gets special treatment in `eval_place`. When `eval_place` and `eval_place_to_op` are inconsistent, all hell breaks loose.

Reply:
Oh and you are probably also running into

https://github.com/rust-lang/rust/blob/f3634503587a1f4e3b34f96452163595f242f911/src/librustc_mir/interpret/eval_context.rs#L610

I suggest grepping the entire `interpret` module for `RETURN_PLACE`.

Reply:
Yeah I did that, but the ICE still happens :/

Reply:
Did what?

In the current state, what you did here `operand.rs` cannot work. You need to look up the return place in the current frame and return that (as an operand).

Reply:
Okay, I think I've fixed it now. I had some other issues in the NRVO pass and thought that this code was the cause.